cpp/src/arrow/dataset/partition.cc

```diff
@@ -410,21 +434,26 @@ std::shared_ptr<PartitioningFactory> DirectoryPartitioning::MakeFactory(
 }
 
 util::optional<KeyValuePartitioning::Key> HivePartitioning::ParseKey(
-    const std::string& segment) {
+    const std::string& segment, const std::string& null_fallback) {
   auto name_end = string_view(segment).find_first_of('=');
+  // Not round-trippable
   if (name_end == string_view::npos) {
     return util::nullopt;
   }
 
-  return Key{segment.substr(0, name_end), segment.substr(name_end + 1)};
+  auto value = segment.substr(name_end + 1);
+  if (value == null_fallback) {
+    return Key{segment.substr(0, name_end), "", true};
+  }
+  return Key{segment.substr(0, name_end), segment.substr(name_end + 1), false};
```

Comment:
```suggestion
  auto name = segment.substr(0, name_end);
  auto value = segment.substr(name_end + 1);
  if (value == null_fallback) {
    return Key{name, "", true};
  }
  return Key{name, value, false};
```

Reply:
Made the change (not through git since there were other changes here)